src/main/clojure/clara/rules/compiler.clj

```diff
@@ -1078,27 +1078,36 @@
                            fact-binding (conj fact-binding))
 
             ;; Find children that all parent nodes have.
-            common-children-ids (into
-                                 ;; Since we require that id-to-condition-nodes have an equal value to "node" under the ID
-                                 ;; for this to be used, the order of this set probably doesn't matter in most cases.  However, in any possible edge
-                                 ;; cases where there are equal nodes under different IDs, sorting the set will add determinism.
-                                 ;; Having different nodes under the same
-                                 ;; ID would be a bug, but having an equivalent node under multiple IDs wouldn't necessarily be one.
-                                 (sorted-set)
-                                 (apply s/union (-> (:forward-edges beta-graph)
-                                                    (select-keys parent-ids)
-                                                    (vals))))
-
+            forward-edges (if (= 1 (count parent-ids))
+                            ;; If there is only one parent then there is no need to reconstruct
+                            ;; the set of forward edges
+                            (-> (:forward-edges beta-graph)
+                                (get (first parent-ids)))
+                            (->> (select-keys parent-ids (:forward-edges beta-graph))
+                                 vals
+                                 (into [] (comp cat (distinct)))))
 
             id-to-condition-nodes (:id-to-condition-node beta-graph)
 
-            ;; Find an id for an equivalent node.
-            existing-id (first (for [child-id common-children-ids
-                                     :when (= node (get id-to-condition-nodes child-id))]
-                                 child-id))
+            ;; Since we require that id-to-condition-nodes have an equal value to "node" under the ID
+            ;; for this to be used, the order of this set probably doesn't matter in most cases.  However, in any possible edge
+            ;; cases where there are equal nodes under different IDs, sorting the set will add determinism.
+            ;; Having different nodes under the same
+            ;; ID would be a bug, but having an equivalent node under multiple IDs wouldn't necessarily be one.
+            update-node->id (fn [m id]
+                              (let [node (get id-to-condition-nodes id)
```

Comment:
Since the node here is part of the id-to-condition-nodes, part of the beta-graph that is built up iteratively, I believe this will the same object if used in multiple iterations of this.  As a result the hashcode will be cached, minimizing the cost.  In the [current version on the other hand](https://github.com/cerner/clara-rules/blob/0.17.0/src/main/clojure/clara/rules/compiler.clj#L1072) the node that we [do an equality check on later](https://github.com/cerner/clara-rules/blob/0.17.0/src/main/clojure/clara/rules/compiler.clj#L1097) is a new object, which limits opportunities to reduce the work.  I suggest mentioning this facilitating of hashcode caching in a comment.